### ISSUE TITLE
Opts and callback for put/del/batch should be optional arguments

### ISSUE THREAD

**Issue by vweevers** [deanlandolt/bytespace#1]
Great work on this; it's *almost* a drop-in replacement for level-sublevel. I was looking to get away from the `level-sublevel/bytewise` requirement in level-scout, and more precisely, to have independent encodings for sublevels. bytespace handles that with grace. The only incompatibilities (with both sublevel and levelup) I've noticed so far are the function signatures of put, del and batch, which should be

- `put(key, value[, options][, callback])`
- `del(key[, options][, callback])`
- `batch(array[, options][, callback])`

PS. bytespace + bytewise key encoding + memdown also *almost* works, I just need to do a tiny PR to memdown that fixes Buffer comparisons :)

**Comment by deanlandolt** [comment]
Thanks @vweevers -- and sorry for taking so long to get back to you (I was traveling this last week).

So I just pushed a new version that fixed the optional args and the one huge issue that was driving me insane -- respecting the `valueEncoding` of the parent db. This should work now (according to the tests at least :) -- I'd be curious to hear if you run into anything around this.

Glad to hear you've nearly got it working with memdown -- that'd be pretty awesome. I'm working through some remaining issues to get it to play nicely with `multilevel` as well, so I can finally create a `level-live-stream` on subspaces without the host db having to know of every sublevel ahead of time -- once I nail that we should be battle-testing this in production in the near future.

**Comment by vweevers** [comment]
Okay, thanks. That [commit](https://github.com/deanlandolt/bytespace/commit/0ca5ce44f87fbfbd7b29c97449daedc33fed0ac6) breaks things for me though, getting a EncodingError from level-iterator-stream. Seems there's double JSON decoding going on, I'll see if I can write a test to replicate it.

**Comment by deanlandolt** [comment]
Strange. I honestly couldn't figure out how it was working for you before.
I actually have tests in there with nested subspaces using different
encoding, but it's very possible I'm just doing it wrong. If you get a
chance, pass along a gist and I'll take a look.

**Comment by vweevers** [comment]
See vweevers/bytespace@9973946

**Comment by deanlandolt** [comment]
Awesome, thanks! I'll get a fix in. I found a cleaner way to do the encodings last night while trying to find the cleanest way to get this to work over `multilevel`: https://github.com/deanlandolt/bytespace/blob/master/proxy.js 

I'm starting to think `level-updown` might be more trouble than it's worth -- and I'll just wrap the provided db methods like I'm doing in the `bytespace/proxy` module. There's more fiddling with getting the args right, but it's a lot easier to reason about.

I'll try to get a fix in today.

**Comment by vweevers** [comment]
Cool, I'm curious to see what you'll come up with. Agree on `level-updown`, also makes debugging hell.

**Comment by deanlandolt** [comment]
Actually, that `bytespace/proxy` module is *almost* ready -- all that's missing is the commit hooks. I'm working on faking them with a live stream for secondary indexing, but that's far from ideal, and doesn't really fit your use case. I could actually just port over the hooks stuff from `level-updown` version and just kill off that version. I may even be able to get away with just adding `level-hooks` to make it work. But none of this will work remotely -- that would require serializing and reviving the entire call graph of hook functions over the rpc stream. Doable -- and AFAICT it's the only reasonable option -- but it'll take some thought.

**Comment by deanlandolt** [comment]
Just pushed a new release -- let me know if you hit any snags.

**Comment by vweevers** [comment]
Found three issues so far with 0.6.0. The prefix in a hook doesn't seem to be honored:

```js
db.pre(function(op, add){
  add({key: 'b', value: 'world', prefix: sub})
})

db.put('a', 'hello', function(){
  db.get('a', function(err, a){
    t.same(a, 'hello')
  })

  sub.get('b', function(err, b){
    t.same(b, 'world')
  })
})
```

It actually inserts `b` into `db`, instead of the sublevel.

Second issue: that `prefix` option should be optional (not an problem for me, but I noticed it by chance).

And the third: bytespace is incompatible with `level-fix-range`, which adds `start` and `end` options even if they weren't specified. Leads to `{start: undefined, end: undefined}`, which Buffer will complain about back in bytespace-land. This is actually an issue with level-fix-range, but if you could account for it in bytespace, will save some time :+1:

**Comment by deanlandolt** [comment]
Apparently I wasn't careful enough when porting the hook/prefix stuff over from the `updown` version. They're not a feature I'm really using yet since they fall down when used over multilevel, so I kinda forgot the coverage was so light. I'll let you know when I have some stronger tests in and fixes in.

Out of curiosity, how blocked are you by this? Is the transition to bytespace something you'd like to land ASAP? If so, I'll try to get it today.

**Comment by vweevers** [comment]
Thanks. No, you can slow down a bit. Wouldn't mind it of course, but no rush. I'm using scout and bytespace in a prototype visualization app, private access, so I'm fine with a couple of crashes and using some temporary fixes. Besides, allows me to focus on other things

**Comment by deanlandolt** [comment]
@vweevers I pushed a some recent fixes and cleanups along with some new tests -- take it for another spin if you have the time. Also, I realized a few days back that I had a few lines of critical code in a local branch when I asked you to kick the tires last time that were pretty important, and may have had something to do with the wheels coming off.

FWIW I'm about to start using the `prefix` feature in batches pretty heavily, so I'll get some solid test coverage in there for it very soon.
